package/src/bin/omicron-package.rs

```diff
@@ -38,6 +40,83 @@ enum SubCommand {
     Deploy(DeployCommand),
 }
 
+#[derive(Clone, Debug)]
+struct Target(BTreeMap<String, String>);
+
+impl Target {
+    // Returns true if this target should include the package.
+    fn includes_package(
+        &self,
+        log: &Logger,
+        pkg_name: &String,
+        pkg: &Package,
+    ) -> bool {
+        let valid_targets = if let Some(targets) = &pkg.only_for_targets {
+            // If targets are specified for the packages, filter them.
+            targets
+        } else {
+            // If no targets are specified, assume the package should be
+            // included by default.
+            return true;
+        };
+
+        // For each of the targets permitted by the package, check if
+        // the current target matches.
+        for (k, v) in valid_targets {
+            let target_value = if let Some(target_value) = self.0.get(k) {
+                target_value
+            } else {
+                info!(log, "{pkg_name}: Omitting package: Missing target value for key {k}");
+                return false;
+            };
+
+            if target_value != v {
+                info!(log, "{pkg_name}: Omitting package: Current target '{target_value}' != {v}");
+                return false;
+            };
+        }
+        info!(log, "{pkg_name}: Including because target matches");
+        return true;
+    }
+}
+
+impl std::fmt::Display for Target {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        for (key, value) in &self.0 {
+            write!(f, "{}={} ", key, value)?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+enum TargetParseError {
+    #[error("Cannot parse key-value pair out of '{0}'")]
+    MissingEquals(String),
+}
+
+fn parse_target(s: &str) -> Result<Target, TargetParseError> {
+    let kvs = s
+        .split_whitespace()
+        .map(|kv| {
+            kv.split_once('=')
+                .ok_or_else(|| TargetParseError::MissingEquals(kv.to_string()))
+                .map(|(k, v)| (k.to_string(), v.to_string()))
+        })
+        .collect::<Result<BTreeMap<String, String>, _>>()?;
+    Ok(Target(kvs))
+}
+
+// Describes the default target.
+//
+// NOTE: This was mostly included so that the addition of the "--target"
+// flag does not break anyone's existing workflow. However, picking a default
+// here is an onerous prospect, and it may make more sense to require user
+// involvement. If we choose to remove it, having users pick one of a few
+// "build profiles" (in other words, a curated list of target strings)
+// seems like a promising alternative.
+const DEFAULT_TARGET: &str = "switch_variant=tofino_stub";
```

Comment:
Maybe I'm missing something, but isn't this going to mean that every package will be installed only in a "tofino_stub" environment?  It seems like you would want the default to be "install in all environments" unless there is a --target assigned.

Reply:
Oh, I am missing something.  This is on the "install package" path, not the "build package" path.

Reply:
- Packages can be labelled with the [`only_for_targets`](https://github.com/oxidecomputer/omicron-package/pull/19) field.
- If a package has that field, it is compared with the "target" supplied as an argument to the packaging tool (this is the case for dendrite now).
- If a package does *not* have that field (see: all the other packages), the assumption is that it should be included unconditionally.

The logic is laid out in code in `Target::includes_package`, which actually does the filtering on which set of packages to operate on:

https://github.com/oxidecomputer/omicron/pull/1759/files#diff-4a9475e577dfcd00783478a110d40ef986d05638d72a372c8987ff36605b742eR48-R80

So, let's suppose we have a couple packages:

```toml
[package.foo]
service_name = "foo"
zone = true

[package.bar]
service_name = "bar"
zone = true
[package.bar.only_for_targets]
bar_variant = "123"

[package.baz]
service_name = "baz"
zone = true
[package.baz.only_for_targets]
baz_variant = "abc"
```

- No target selects "foo" only
- `--target "bar_variant=123"` selects "foo" + "bar", but not baz
- `--target "bar_variant=123 baz_variant=abc"` selects "foo" + "bar" + "baz"
- `--target "bar_variant=456 baz_variant=def"` selects "foo" only (yeah, it has the keys of the packages we care about, but the values don't match).